modules/deployment-scripts/create-agw-kv-certificate/main.bicep

```diff
@@ -0,0 +1,229 @@
+  name: akvName
+}
+
+resource agw 'Microsoft.Network/applicationGateways@2021-05-01' existing = if (!empty(agwName) && agwCertType != 'none') {
+  name: agwName
+}
+
+@description('The User Assigned Identity of the Azure Application Gateway')
+param agwIdName string = ''
+resource agwId 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' existing = if (!empty(agwName) && !empty(agwIdName) && agwCertType != 'none') {
+  name: agwIdName
+}
+
+@description('A new managed identity that will be created in this Resource Group, this is the default option')
+resource newDepScriptId 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' = if (!useExistingManagedIdentity) {
+  name: managedIdentityName
+  location: location
+}
+
+@description('An existing managed identity that could exist in another sub/rg')
+resource existingDepScriptId 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' existing = if (useExistingManagedIdentity ) {
+  name: managedIdentityName
+  scope: resourceGroup(existingManagedIdentitySubId, existingManagedIdentityResourceGroupName)
+}
+
+resource rbacKv 'Microsoft.Authorization/roleAssignments@2020-08-01-preview' = [for roleDefId in rbacRolesNeededOnKV: {
+  name: guid(akv.id, roleDefId, useExistingManagedIdentity ? existingDepScriptId.id : newDepScriptId.id)
+  scope: akv
+  properties: {
+    roleDefinitionId: resourceId('Microsoft.Authorization/roleDefinitions', roleDefId)
+    principalId: useExistingManagedIdentity ? existingDepScriptId.properties.principalId : newDepScriptId.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}]
+
+resource rbacAppGw 'Microsoft.Authorization/roleAssignments@2020-08-01-preview'  =  [for roleDefId in rbacRolesNeededOnAppGw: if (!empty(agwName) && agwCertType != 'none') {
+  name: guid(agw.id, roleDefId, useExistingManagedIdentity ? existingDepScriptId.id : newDepScriptId.id)
+  scope: agw
+  properties: {
+    roleDefinitionId: resourceId('Microsoft.Authorization/roleDefinitions', roleDefId)
+    principalId: useExistingManagedIdentity ? existingDepScriptId.properties.principalId : newDepScriptId.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}]
+
+var managedIdentityOperator = resourceId('Microsoft.Authorization/roleDefinitions', 'f1a07417-d97a-45cb-824c-7a7467783830')
+@description('Managed identity requires "Managed Identity Operator" permission over the user assigned identity of the Application Gateway.')
+resource agwIdRbac 'Microsoft.Authorization/roleAssignments@2020-08-01-preview' =  if (!empty(agwName) && agwCertType != 'none') {
+  scope: agwId
+  name: guid(agwId.id, managedIdentityOperator, useExistingManagedIdentity ? existingDepScriptId.id : newDepScriptId.id)
+  properties: {
+    roleDefinitionId: managedIdentityOperator
+    principalId: useExistingManagedIdentity ? existingDepScriptId.properties.principalId : newDepScriptId.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}
+
+@batchSize(1)
```

Comment:
why 1?  Also, wouldn't it make more sense to loop in the script itself?

Reply:
AppGw seems to have concurrency problems with multiple updates at the same time. Batchsize was added to mitigate this